File: solutions-by-language/cpp/2197_replace_non_coprime_numbers_in_array.cpp
class Solution {
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        int n = nums.size();
        vector<int> st;
        st.reserve(n);

        for(int x : nums) {
            int cur = x;
            while (!st.empty()) {
                int a = st.back();
                int g = gcd(a,cur);
                if(g == 1) break;
                st.pop_back();
                cur = (a / g ) * cur;
            }
            st.push_back(cur);
        }
        return st;
    }

private:
    static int gcd(int a , int b){
        while(b) {
            int temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }
};

File: solutions-by-language/python/2197_replace_non_coprime_numbers_in_array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack: List[int] = []
        for x in nums:
            cur = x
            while stack:
                a = stack[-1]
                g = math.gcd(a, cur)
                if g == 1:
                    break
                stack.pop()
                cur = (a // g) * cur
            stack.append(cur)
        return stack